### ISSUE TITLE
Citekey contains ":" character

### ISSUE THREAD

**Issue by santtu** [retorquere/zotero-better-bibtex#1553]
**Report ID:** QQA6BFZW-euc

**Expected behavior:** In previous version (not sure which, but prior to 5.2.31) the citekey was `ren_vapor_2019` and I would expect it to stay the same

**Actual behavior:** Upon regeneration of the BIB file, the key changed to `ren_vapor:_2019` (notice extra ":" character)

The article in question is this one: https://link.springer.com/chapter/10.1007%2F978-3-030-32101-7_29 and the full title is "VAPOR: A Value-Centric Blockchain that is Scale-out, Decentralized, and Flexible by Design".

Curiously enough none of the other 86 citations in this collection changed they citekey. I tried unpinning the key and refreshing it, but the colon character persisted.

The last time I can accurately place a `ren_vapor_2019` citekey being generated is on April 30th, sometime between that and now the key changed to one with colon.

I noticed this with 5.2.31, and upgraded to 5.2.32 which showed the same behavior.

(As a workaround, I pinned the old citekey via extra field...)

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.2.32.6691](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.32.6691.xpi) ("zotero pattern changed behavior")

Install in Zotero by downloading [test build 5.2.32.6691](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.32.6691.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by retorquere** [comment]
I'm going to take a wild guess and say that 

> VAPOR: A Value-Centric Blockchain that is Scale-out, Decentralized, and Flexible by Design

has been in your library since before 2020-01-01?

You have as your pattern `[zotero:clean]`, and `zotero` uses the Zotero citekey generator (AAMOF it lifts the algorithm from the Zotero BibTeX exporter and uses it as-is), and that generator distinguishes between items that had a `dateAdded` with a year of `2020` or later, and anything else. For the "older" items, the Zotero algorithm retains a-z , 0-9, and "!$&\*+-./:;\>?[]^\_\`|". For the others, it retains a-z, 0-9, `_` and `-`. The rationale for this is given as:

>  // use legacy pattern for all old items to not break existing usages
>  // but use the simple pattern for all newly added items
>  // or always if the hiddenPref is set
>  // extensions.zotero.translators.BibTeX.export.simpleCitekey

You can thus create the [hidden preference](https://www.zotero.org/support/preferences/hidden_preferences) `extensions.zotero.translators.BibTeX.export.simpleCitekey` and set it to `true`, and that will always treat items as >= 2020. There was a bug in my lift-the-algorith-from-zotero that would not read the preference -- 6691 fixes that. 
But more generally, what you get with `[zotero]` is just what Zotero itself would generate for the item. If Zotero changes, so will `[zotero]`.

**Comment by santtu** [comment]
The article was imported on October 10th 2019 (last year). I imported another version of the same article, and it got a clean citekey (`ren_vapor_2019-1`). 

However, this still begs the question why an earlier version of Zotero generated citekey `ren_vapor_2019` and a newer version generated `ren_vapor:_2019`? Why didn't it contain colon in the the earlier export? Was this some change in Zotero or Zotero BibTeX exporter between then and now? This sounds like a regression **somewhere**, but what would then be the correct place to report it?

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5.2.32.6693](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.32.6693.xpi) ("prefs cleanup")

Install in Zotero by downloading [test build 5.2.32.6693](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.2.32.6693.xpi), opening the Zotero "Tools" menu, selecting "Add-ons", open the gear menu in the top right, and select "Install Add-on From File...".

**Comment by retorquere** [comment]
It remains to be answered, rather 😉. But you are right; I had forgotten that I made a recent change that *also* affected this; `clean` is supposed to only remove invalid characters, and it would previously also remove `:`, but `:` is not invalid in a key. I'm conflicted about what to do about this -- you are right that this is a behavior change, but some people do want these colon characters in their keys.

**Comment by santtu** [comment]
So... let me rephrase and see if I got this correct or not.

1. Zotero citekey generator would generate `ren_vapor:_2019`
2. Better BibTeX prior to 6991 incorrectly read `simpleCiteKey` and operated as if it was always `true`
3. This meant that it was generating clean keys, which resulted in `ren_vapor_2019`
4. In 6991, `simpleCiteKey` is now correctly read and handled
5. And since I **did not have it set**, it was handled as `false` value
6. Consequently the citekey now included the "non-clean" character i.e. `ren_vapor:_2019`

So this isn't really a bug in Zotero (since it, by definition, allows colons in citekey and nothing has changed there), and not ... really a bug in Better BibTeX ... it was just that my bibliography was actually unintentionally relying on a bug generating a citekey in format that was based on a bug that handled `simpleCiteKey` incorrectly ... is that it?

**Comment by retorquere** [comment]
> So... let me rephrase and see if I got this correct or not.

Almost

>     1. Zotero citekey generator would generate `ren_vapor:_2019`

Correct (for you -- when I imported it, it would have today's date for the item)

>     2. Better BibTeX prior to 6991 incorrectly read `simpleCiteKey` and operated as if it was always `true`

As if it was always `false`. I couldn't read the pref correctly, and then it defaults to `false` in Zotero. So my first diagnosis was wrong for your particular case (but I did fix BBT to read the pref)

>     3. This meant that it was generating clean keys, which resulted in `ren_vapor_2019`

It was actually generating ren_vapor:_2019, because of the date check, and that `simpleCitekey` wasn't on as far as BBT knew.

>     4. In 6991, `simpleCiteKey` is now correctly read and handled

That is correct.

>     5. And since I **did not have it set**, it was handled as `false` value

That is also correct

>     6. Consequently the citekey now included the "non-clean" character i.e. `ren_vapor:_2019`

And here is where my original diagnosis was wrong. The behavior I described was correct, but that meant it was always generating `ren_vapor:_2019`. It's just that `clean` removed `:` until recently, which it shouldn't have given what I had in mind for it to do.

> So this isn't really a bug in Zotero (since it, by definition, allows colons in citekey 

It's more that bibtex toolchains allow colons in the citekey.  Zotero (and thus `[zotero]`) could generate citekeys that bibtex will not accept. I reckon this is wht they introduce `simpleCiteKey`. But no, not a bug in Zotero -- there can be no bug in Zotero wrt this, because by design, `[zotero]` should do what Zotero does, warts and all (with "warts" being mostly pre-`simpleCiteKey`).

> and nothing has changed there), 

correct.

> and not ... really a bug in Better BibTeX ... 

Well... this is where it gets murky. For starters there was a bug on `simpleCiteKey` handling. But then there was also the behavior change in `clean`.

> it was just that my bibliography was actually unintentionally relying on a bug generating a citekey in format that was based on a bug that handled `simpleCiteKey` incorrectly ... is that it?

That was what I thought it was, and it *was* what was happening, but you'd get the citekey with the `:`; it just wasn't possible to turn on `simpleCiteKey` as far as BBT could tell. What you saw was that `clean` would previously remove the `:` where it shouldn't have, given what it's purpose is. I'm now torn between keeping a fix for what I consider to be a flaw (`clean` should never have removed the `:`) and retaining existing behavior.

In the course of this I also found *another* bug in a filter; `nopunct` would change `_` to `-`. 6693 fixes that and that would make `[zotero:clean:nopunct]` (or `[zotero:nopunct:clean]`, which does the same in this case) an additional way to get what you expect.

**Comment by santtu** [comment]
This got complicated really fast :-) However, from my point of view I understand why the thing happened, I have a workaround on that. FWIW, this is closeable. It is your call to decide whether this is a side effect of a bug fix, or a bug in itself, and how to continue from there.

**Comment by retorquere** [comment]
Well you stumbled upon an actual bug :)

I'm going to merge 6693, and then either `simpleCiteKey` or `nopunct` will work.

You can keep 6693, you will be upgraded to the new release automatically.
